Fix indent v2

cmrschwarz opened the pull request:
This is my second attempt at fixing the more involved indenting issues first brought up in #650 .

It turns out that the `Indent` node is simply not powerful enough to handle all cases :(.

We will unfortunately have to dynamically track whether a partial ended on a newline or not.
This PR does exactly that.

Here are some of the more difficult edge cases that this solves:

### Input
```
{{#*inline "empty_partial"}}{{/inline}}
<div>
    {{> empty_partial}}
</div>
```

### Output
```
<div>
</div>
``` 
---


### Input
```
{{#*inline "partial"}}foo{{/inline}}
<div>
    {{> partial}}
</div>
```

### Output 
```
<div>
    foo</div>
```
(yes, this is indeed correct according to [handlebars js](https://handlebarsjs.com/playground.html#format=1&currentExample=%7B%22template%22%3A%22%7B%7B%23*inline%20%5C%22partial%5C%22%7D%7Dfoo%7B%7B%2Finline%7D%7D%5Cn%3Cdiv%3E%5Cn%20%20%20%20%7B%7B%3E%20partial%7D%7D%5Cn%3C%2Fdiv%3E%22%2C%22partials%22%3A%5B%5D%2C%22input%22%3A%22%7B%7D%5Cn%22%2C%22output%22%3A%22%5Cn%3Cdiv%3E%5Cn%20%20%20%20foo%3C%2Fdiv%3E%22%2C%22preparationScript%22%3A%22%22%2C%22handlebarsVersion%22%3A%224.7.8%22%7D))

---
### Input
Json Input: `{"a": "line\ndynamic_trailing_newline\n"}`
```
{{#*inline "dynamic_partial"}}{{a}}{{/inline}}
<div>
    {{> dynamic_partial}}
</div>
```
### Output
```
<div>
    line
    dynamic_trailing_newline
</div>
```

sunng87 commented on src/template.rs:
```diff
@@ -143,10 +144,11 @@ pub struct HelperTemplate {
     pub inverse: Option<Template>,
     pub block: bool,
     pub chain: bool,
+    pub indent_before_write: bool,
```
I think we can make this field private, as well as the one for `DecoratorTemplate`

cmrschwarz replied:
Sure!  I need them in `render_helper`, so I guess `pub(crate)` it is.

sunng87 reviewed (commented):
I have been super busy these days. Really need some time to deep dive into this. But the test cases look good. I think this approach is more promising.

cmrschwarz commented:
I've been using this in practice for a bit more now and found two minor issues with it.
The commits just now fixed those and added regression tests.
Otherwise this has worked nicely, indenting seems just about perfect now.

sunng87 commented:
This looks good to me. Thank you for the fix! I have been struggling with those indent for quite a while. This approach seems practical and also keeps complexity at a acceptable level.

sunng87 commented:
I just tried the case reported in #611, and it seems we still this corner case to cover.

cmrschwarz commented:
> I just tried the case reported in #611, and it seems we still this corner case to cover.

 I'm happy to look into that, though I'm quite busy this week so it might have to wait until the weekend.

sunng87 commented:
Thank you and no rush @cmrschwarz . Just take your time.
